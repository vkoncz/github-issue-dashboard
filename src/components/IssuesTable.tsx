import { useGithubIssuesQuery } from '../graphql-client/github-issues.query';

export function IssuesTable() {
  const { loading, data } = useGithubIssuesQuery();

  return <p>{!loading && JSON.stringify(data)}</p>;
}
