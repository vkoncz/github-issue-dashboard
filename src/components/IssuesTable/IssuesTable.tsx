import type { ColDef, IDatasource } from 'ag-grid-community';
import { AgGridReact } from 'ag-grid-react';
import { useMemo, useRef } from 'react';

import type { Node } from '../../graphql-client/github-client.model';
import { useGithubIssuesQuery } from '../../graphql-client/github-issues.query';
import { AuthorCell } from '../CustomCells/AuthorCell';
import { IssueLinkCell } from '../CustomCells/IssueLinkCell';
import styles from './IssueTable.module.css';

type IssueColDef = Omit<ColDef, 'field'> & { field: keyof Node };

const columnDefs: IssueColDef[] = [
  { field: 'title', flex: 1 },
  { field: 'author', cellRenderer: AuthorCell },
  { field: 'state' },
  { field: 'createdAt' },
  { field: 'url', cellRenderer: IssueLinkCell },
];

const numberOfLines = 30;

export function IssuesTable() {
  const cursor = useRef<undefined | string>();
  const [getIssues] = useGithubIssuesQuery({});

  const datasource = useMemo<IDatasource>(
    () => ({
      getRows: async params => {
        const result = await getIssues({
          variables: { afterCursor: cursor.current, blockSize: numberOfLines },
        });

        const issues = result.data?.repository.issues;
        if (!issues) return;

        const lastRow = issues.pageInfo.endCursor ? undefined : issues.totalCount;
        params.successCallback(issues.nodes, lastRow);
        cursor.current = issues.pageInfo.endCursor;
      },
    }),
    [getIssues],
  );

  return (
    <div className={`${styles.tableContainer} ag-theme-alpine`}>
      <AgGridReact
        columnDefs={columnDefs}
        rowBuffer={0}
        gridOptions={{
          rowModelType: 'infinite',
          datasource,
          cacheBlockSize: numberOfLines,
        }}
      />
    </div>
  );
}
