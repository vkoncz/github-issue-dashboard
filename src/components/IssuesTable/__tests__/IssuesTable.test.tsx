import { render } from '@testing-library/react';

import { IssuesTable } from '../IssuesTable';

jest.mock('../../../graphql-client/github-issues.query', () => ({
  useGithubIssuesQuery: () => [() => {}],
}));

describe('IssueTable', () => {
  it('should match snapshot', () => {
    const { container } = render(<IssuesTable />);

    expect(container).toMatchSnapshot();
  });
});
