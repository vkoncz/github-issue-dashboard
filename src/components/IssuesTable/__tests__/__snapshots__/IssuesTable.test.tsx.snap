// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueTable should match snapshot 1`] = `
<div>
  <div
    class="tableContainer ag-theme-alpine"
  >
    <div
      style="height: 100%;"
    >
      <!-- AG Grid -->
      <div
        class="ag-root-wrapper ag-ltr ag-layout-normal"
      >
        <div
          class="ag-root-wrapper-body ag-focus-managed ag-layout-normal"
        >
          <div
            class="ag-tab-guard ag-tab-guard-top"
            role="presentation"
            tabindex="0"
          />
          <!-- AG Grid Body -->
          <div
            aria-colcount="5"
            aria-rowcount="-1"
            class="ag-root ag-unselectable ag-layout-normal"
            role="grid"
          >
            <div
              class="ag-header ag-pivot-off"
              role="presentation"
              style="height: 49px; min-height: 49px;"
            >
              <div
                class="ag-pinned-left-header ag-hidden"
                role="presentation"
                style="width: 0px; min-width: 0px; max-width: 0px;"
              >
                <div
                  aria-rowindex="1"
                  class="ag-header-row ag-header-row-column"
                  role="row"
                  style="height: 48px; top: 0px; width: 0px;"
                />
              </div>
              <div
                class="ag-header-viewport "
                role="presentation"
              >
                <div
                  class="ag-header-container"
                  role="rowgroup"
                  style="width: 1000px;"
                >
                  <div
                    aria-rowindex="1"
                    class="ag-header-row ag-header-row-column"
                    role="row"
                    style="height: 48px; top: 0px; width: 1000px;"
                  >
                    <div
                      aria-colindex="1"
                      class="ag-header-cell ag-focus-managed"
                      col-id="title"
                      role="columnheader"
                      style="left: 0px; width: 200px;"
                      tabindex="-1"
                    >
                      <div
                        class="ag-header-cell-resize ag-hidden"
                        role="presentation"
                      />
                      <div
                        class="ag-labeled ag-label-align-right ag-checkbox ag-input-field ag-header-select-all ag-hidden"
                        role="presentation"
                      >
                        
                
                        <div
                          class="ag-input-field-label ag-label ag-hidden ag-checkbox-label"
                          ref="eLabel"
                          role="presentation"
                        />
                        
                
                        <div
                          class="ag-wrapper ag-input-wrapper ag-checkbox-input-wrapper"
                          ref="eWrapper"
                          role="presentation"
                        >
                          
                    
                          <input
                            aria-label="Press Space to toggle all rows selection (unchecked)"
                            class="ag-input-field-input ag-checkbox-input"
                            id="ag-3-input"
                            ref="eInput"
                            tabindex="-1"
                            type="checkbox"
                          />
                          
                
                        </div>
                        
            
                      </div>
                      <div
                        class="ag-cell-label-container"
                        role="presentation"
                      >
                        
            
                        
            
                        <div
                          class="ag-header-cell-label"
                          ref="eLabel"
                          role="presentation"
                        >
                          
                
                          <span
                            class="ag-header-cell-text"
                            ref="eText"
                          >
                            Title
                          </span>
                          
                
                          <span
                            aria-hidden="true"
                            class="ag-header-icon ag-header-label-icon ag-filter-icon ag-hidden"
                            ref="eFilter"
                          >
                            <span
                              class="ag-icon ag-icon-filter"
                              role="presentation"
                              unselectable="on"
                            />
                          </span>
                          
                
                          
                
                          
                
                          
                
                          
            
                        </div>
                        
        
                      </div>
                    </div>
                    <div
                      aria-colindex="2"
                      class="ag-header-cell ag-focus-managed"
                      col-id="author"
                      role="columnheader"
                      style="left: 200px; width: 200px;"
                      tabindex="-1"
                    >
                      <div
                        class="ag-header-cell-resize ag-hidden"
                        role="presentation"
                      />
                      <div
                        class="ag-labeled ag-label-align-right ag-checkbox ag-input-field ag-header-select-all ag-hidden"
                        role="presentation"
                      >
                        
                
                        <div
                          class="ag-input-field-label ag-label ag-hidden ag-checkbox-label"
                          ref="eLabel"
                          role="presentation"
                        />
                        
                
                        <div
                          class="ag-wrapper ag-input-wrapper ag-checkbox-input-wrapper"
                          ref="eWrapper"
                          role="presentation"
                        >
                          
                    
                          <input
                            aria-label="Press Space to toggle all rows selection (unchecked)"
                            class="ag-input-field-input ag-checkbox-input"
                            id="ag-4-input"
                            ref="eInput"
                            tabindex="-1"
                            type="checkbox"
                          />
                          
                
                        </div>
                        
            
                      </div>
                      <div
                        class="ag-cell-label-container"
                        role="presentation"
                      >
                        
            
                        
            
                        <div
                          class="ag-header-cell-label"
                          ref="eLabel"
                          role="presentation"
                        >
                          
                
                          <span
                            class="ag-header-cell-text"
                            ref="eText"
                          >
                            Author
                          </span>
                          
                
                          <span
                            aria-hidden="true"
                            class="ag-header-icon ag-header-label-icon ag-filter-icon ag-hidden"
                            ref="eFilter"
                          >
                            <span
                              class="ag-icon ag-icon-filter"
                              role="presentation"
                              unselectable="on"
                            />
                          </span>
                          
                
                          
                
                          
                
                          
                
                          
            
                        </div>
                        
        
                      </div>
                    </div>
                    <div
                      aria-colindex="3"
                      class="ag-header-cell ag-focus-managed"
                      col-id="state"
                      role="columnheader"
                      style="left: 400px; width: 200px;"
                      tabindex="-1"
                    >
                      <div
                        class="ag-header-cell-resize ag-hidden"
                        role="presentation"
                      />
                      <div
                        class="ag-labeled ag-label-align-right ag-checkbox ag-input-field ag-header-select-all ag-hidden"
                        role="presentation"
                      >
                        
                
                        <div
                          class="ag-input-field-label ag-label ag-hidden ag-checkbox-label"
                          ref="eLabel"
                          role="presentation"
                        />
                        
                
                        <div
                          class="ag-wrapper ag-input-wrapper ag-checkbox-input-wrapper"
                          ref="eWrapper"
                          role="presentation"
                        >
                          
                    
                          <input
                            aria-label="Press Space to toggle all rows selection (unchecked)"
                            class="ag-input-field-input ag-checkbox-input"
                            id="ag-5-input"
                            ref="eInput"
                            tabindex="-1"
                            type="checkbox"
                          />
                          
                
                        </div>
                        
            
                      </div>
                      <div
                        class="ag-cell-label-container"
                        role="presentation"
                      >
                        
            
                        
            
                        <div
                          class="ag-header-cell-label"
                          ref="eLabel"
                          role="presentation"
                        >
                          
                
                          <span
                            class="ag-header-cell-text"
                            ref="eText"
                          >
                            State
                          </span>
                          
                
                          <span
                            aria-hidden="true"
                            class="ag-header-icon ag-header-label-icon ag-filter-icon ag-hidden"
                            ref="eFilter"
                          >
                            <span
                              class="ag-icon ag-icon-filter"
                              role="presentation"
                              unselectable="on"
                            />
                          </span>
                          
                
                          
                
                          
                
                          
                
                          
            
                        </div>
                        
        
                      </div>
                    </div>
                    <div
                      aria-colindex="4"
                      class="ag-header-cell ag-focus-managed"
                      col-id="createdAt"
                      role="columnheader"
                      style="left: 600px; width: 200px;"
                      tabindex="-1"
                    >
                      <div
                        class="ag-header-cell-resize ag-hidden"
                        role="presentation"
                      />
                      <div
                        class="ag-labeled ag-label-align-right ag-checkbox ag-input-field ag-header-select-all ag-hidden"
                        role="presentation"
                      >
                        
                
                        <div
                          class="ag-input-field-label ag-label ag-hidden ag-checkbox-label"
                          ref="eLabel"
                          role="presentation"
                        />
                        
                
                        <div
                          class="ag-wrapper ag-input-wrapper ag-checkbox-input-wrapper"
                          ref="eWrapper"
                          role="presentation"
                        >
                          
                    
                          <input
                            aria-label="Press Space to toggle all rows selection (unchecked)"
                            class="ag-input-field-input ag-checkbox-input"
                            id="ag-6-input"
                            ref="eInput"
                            tabindex="-1"
                            type="checkbox"
                          />
                          
                
                        </div>
                        
            
                      </div>
                      <div
                        class="ag-cell-label-container"
                        role="presentation"
                      >
                        
            
                        
            
                        <div
                          class="ag-header-cell-label"
                          ref="eLabel"
                          role="presentation"
                        >
                          
                
                          <span
                            class="ag-header-cell-text"
                            ref="eText"
                          >
                            Created At
                          </span>
                          
                
                          <span
                            aria-hidden="true"
                            class="ag-header-icon ag-header-label-icon ag-filter-icon ag-hidden"
                            ref="eFilter"
                          >
                            <span
                              class="ag-icon ag-icon-filter"
                              role="presentation"
                              unselectable="on"
                            />
                          </span>
                          
                
                          
                
                          
                
                          
                
                          
            
                        </div>
                        
        
                      </div>
                    </div>
                    <div
                      aria-colindex="5"
                      class="ag-header-cell ag-focus-managed"
                      col-id="url"
                      role="columnheader"
                      style="left: 800px; width: 200px;"
                      tabindex="-1"
                    >
                      <div
                        class="ag-header-cell-resize ag-hidden"
                        role="presentation"
                      />
                      <div
                        class="ag-labeled ag-label-align-right ag-checkbox ag-input-field ag-header-select-all ag-hidden"
                        role="presentation"
                      >
                        
                
                        <div
                          class="ag-input-field-label ag-label ag-hidden ag-checkbox-label"
                          ref="eLabel"
                          role="presentation"
                        />
                        
                
                        <div
                          class="ag-wrapper ag-input-wrapper ag-checkbox-input-wrapper"
                          ref="eWrapper"
                          role="presentation"
                        >
                          
                    
                          <input
                            aria-label="Press Space to toggle all rows selection (unchecked)"
                            class="ag-input-field-input ag-checkbox-input"
                            id="ag-7-input"
                            ref="eInput"
                            tabindex="-1"
                            type="checkbox"
                          />
                          
                
                        </div>
                        
            
                      </div>
                      <div
                        class="ag-cell-label-container"
                        role="presentation"
                      >
                        
            
                        
            
                        <div
                          class="ag-header-cell-label"
                          ref="eLabel"
                          role="presentation"
                        >
                          
                
                          <span
                            class="ag-header-cell-text"
                            ref="eText"
                          >
                            Url
                          </span>
                          
                
                          <span
                            aria-hidden="true"
                            class="ag-header-icon ag-header-label-icon ag-filter-icon ag-hidden"
                            ref="eFilter"
                          >
                            <span
                              class="ag-icon ag-icon-filter"
                              role="presentation"
                              unselectable="on"
                            />
                          </span>
                          
                
                          
                
                          
                
                          
                
                          
            
                        </div>
                        
        
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="ag-pinned-right-header ag-hidden"
                role="presentation"
                style="width: 0px; min-width: 0px; max-width: 0px;"
              >
                <div
                  aria-rowindex="1"
                  class="ag-header-row ag-header-row-column"
                  role="row"
                  style="height: 48px; top: 0px; width: 0px;"
                />
              </div>
            </div>
            <!-- AG Pinned Top -->
            <div
              class="ag-floating-top"
              role="presentation"
              style="height: 0px; min-height: 0; display: none;"
            >
              <!-- AG Row Container topLeft -->
              <div
                class="ag-pinned-left-floating-top ag-hidden"
                role="rowgroup"
              />
              <!-- AG Row Container topCenter -->
              <div
                class="ag-floating-top-viewport"
                role="presentation"
              >
                <div
                  class="ag-floating-top-container"
                  role="rowgroup"
                  style="width: 1000px;"
                />
              </div>
              <!-- AG Row Container topRight -->
              <div
                class="ag-pinned-right-floating-top ag-hidden"
                role="rowgroup"
              />
              <!-- AG Row Container topFullWidth -->
              <div
                class="ag-floating-top-full-width-container"
                role="rowgroup"
              />
            </div>
            <!-- AG Middle -->
            <div
              class="ag-body-viewport ag-row-no-animation ag-layout-normal"
              role="presentation"
            >
              <!-- AG Row Container left -->
              <div
                class="ag-pinned-left-cols-container ag-hidden"
                role="rowgroup"
                style="height: 42px;"
              >
                <div
                  aria-rowindex="2"
                  class="ag-row-even ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute ag-row-first ag-row-last"
                  role="row"
                  row-index="0"
                  style="transform: translateY(0px); height: 42px;"
                />
              </div>
              <!-- AG Row Container center -->
              <div
                class="ag-center-cols-clipper"
                role="presentation"
                style="height: 42px;"
              >
                <div
                  class="ag-center-cols-viewport"
                  role="presentation"
                  style="height: 100%;"
                >
                  <div
                    class="ag-center-cols-container"
                    role="rowgroup"
                    style="height: 42px; width: 1000px;"
                  >
                    <div
                      aria-rowindex="2"
                      class="ag-row-even ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute ag-row-first ag-row-last"
                      role="row"
                      row-index="0"
                      style="transform: translateY(0px); height: 42px;"
                    >
                      <div
                        aria-colindex="1"
                        class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
                        col-id="title"
                        role="gridcell"
                        style="left: 0px; width: 200px;"
                        tabindex="-1"
                      />
                      <div
                        aria-colindex="2"
                        class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
                        col-id="author"
                        role="gridcell"
                        style="left: 200px; width: 200px;"
                        tabindex="-1"
                      >
                        <span>
                          <a
                            rel="noreferrer"
                            target="_blank"
                          />
                        </span>
                      </div>
                      <div
                        aria-colindex="3"
                        class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
                        col-id="state"
                        role="gridcell"
                        style="left: 400px; width: 200px;"
                        tabindex="-1"
                      />
                      <div
                        aria-colindex="4"
                        class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
                        col-id="createdAt"
                        role="gridcell"
                        style="left: 600px; width: 200px;"
                        tabindex="-1"
                      />
                      <div
                        aria-colindex="5"
                        class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
                        col-id="url"
                        role="gridcell"
                        style="left: 800px; width: 200px;"
                        tabindex="-1"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <!-- AG Row Container right -->
              <div
                class="ag-pinned-right-cols-container ag-hidden"
                role="rowgroup"
                style="height: 42px;"
              >
                <div
                  aria-rowindex="2"
                  class="ag-row-even ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute ag-row-first ag-row-last"
                  role="row"
                  row-index="0"
                  style="transform: translateY(0px); height: 42px;"
                />
              </div>
              <!-- AG Row Container fullWidth -->
              <div
                class="ag-full-width-container"
                role="rowgroup"
                style="height: 42px;"
              />
            </div>
            <!-- AG Pinned Bottom -->
            <div
              class="ag-floating-bottom"
              role="presentation"
              style="height: 0px; min-height: 0; display: none;"
            >
              <!-- AG Row Container bottomLeft -->
              <div
                class="ag-pinned-left-floating-bottom ag-hidden"
                role="rowgroup"
              />
              <!-- AG Row Container bottomCenter -->
              <div
                class="ag-floating-bottom-viewport"
                role="presentation"
              >
                <div
                  class="ag-floating-bottom-container"
                  role="rowgroup"
                  style="width: 1000px;"
                />
              </div>
              <!-- AG Row Container bottomRight -->
              <div
                class="ag-pinned-right-floating-bottom ag-hidden"
                role="rowgroup"
              />
              <!-- AG Row Container bottomFullWidth -->
              <div
                class="ag-floating-bottom-full-width-container"
                role="rowgroup"
              />
            </div>
            <!-- AG Fake Horizontal Scroll -->
            <div
              aria-hidden="true"
              class="ag-body-horizontal-scroll"
              style="height: 0px; max-height: 0px; min-height: 0px;"
            >
              
            
              <div
                class="ag-horizontal-left-spacer"
                ref="eLeftSpacer"
                style="width: 0px; max-width: 0px; min-width: 0px;"
              />
              
            
              <div
                class="ag-body-horizontal-scroll-viewport"
                ref="eViewport"
                style="height: 0px; max-height: 0px; min-height: 0px;"
              >
                
                
                <div
                  class="ag-body-horizontal-scroll-container"
                  ref="eContainer"
                  style="height: 0px; max-height: 0px; min-height: 0px; width: 1000px;"
                />
                
            
              </div>
              
            
              <div
                class="ag-horizontal-right-spacer"
                ref="eRightSpacer"
                style="width: 0px; max-width: 0px; min-width: 0px;"
              />
              
        
            </div>
            <!-- AG Overlay Wrapper -->
            <div
              aria-hidden="true"
              class="ag-overlay ag-hidden"
            >
              
            
              <div
                class="ag-overlay-panel"
              >
                
                
                <div
                  class="ag-overlay-wrapper ag-layout-normal"
                  ref="eOverlayWrapper"
                />
                
            
              </div>
              
        
            </div>
          </div>
          <div
            class="ag-tab-guard ag-tab-guard-bottom"
            role="presentation"
            tabindex="0"
          />
        </div>
        <div
          class="ag-paging-panel ag-unselectable ag-hidden"
          id="ag-2"
        >
          
                
          <span
            class="ag-paging-row-summary-panel"
            role="status"
          >
            
                    
            <span
              class="ag-paging-row-summary-panel-number"
              id="ag-2-first-row"
              ref="lbFirstRowOnPage"
            />
            
                    
            <span
              id="ag-2-to"
            >
              to
            </span>
            
                    
            <span
              class="ag-paging-row-summary-panel-number"
              id="ag-2-last-row"
              ref="lbLastRowOnPage"
            />
            
                    
            <span
              id="ag-2-of"
            >
              of
            </span>
            
                    
            <span
              class="ag-paging-row-summary-panel-number"
              id="ag-2-row-count"
              ref="lbRecordCount"
            />
            
                
          </span>
          
                
          <span
            class="ag-paging-page-summary-panel"
            role="presentation"
          >
            
                    
            <div
              aria-label="First Page"
              class="ag-paging-button"
              ref="btFirst"
              role="button"
              tabindex="0"
            >
              <span
                class="ag-icon ag-icon-first"
                role="presentation"
                unselectable="on"
              />
            </div>
            
                    
            <div
              aria-label="Previous Page"
              class="ag-paging-button"
              ref="btPrevious"
              role="button"
              tabindex="0"
            >
              <span
                class="ag-icon ag-icon-previous"
                role="presentation"
                unselectable="on"
              />
            </div>
            
                    
            <span
              class="ag-paging-description"
              role="status"
            >
              
                        
              <span
                id="ag-2-start-page"
              >
                Page
              </span>
              
                        
              <span
                class="ag-paging-number"
                id="ag-2-start-page-number"
                ref="lbCurrent"
              />
              
                        
              <span
                id="ag-2-of-page"
              >
                of
              </span>
              
                        
              <span
                class="ag-paging-number"
                id="ag-2-of-page-number"
                ref="lbTotal"
              />
              
                    
            </span>
            
                    
            <div
              aria-label="Next Page"
              class="ag-paging-button"
              ref="btNext"
              role="button"
              tabindex="0"
            >
              <span
                class="ag-icon ag-icon-next"
                role="presentation"
                unselectable="on"
              />
            </div>
            
                    
            <div
              aria-label="Last Page"
              class="ag-paging-button"
              ref="btLast"
              role="button"
              tabindex="0"
            >
              <span
                class="ag-icon ag-icon-last"
                role="presentation"
                unselectable="on"
              />
            </div>
            
                
          </span>
          
            
        </div>
      </div>
    </div>
  </div>
</div>
`;
